tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +429,275 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+    // accept any number of wasted thread and ignore its cost.
+    if (num_full_tiles <= 1) cost = wasted_threads;
+
+    // Using less and equal here because given the same cost, we would like to
+    // launch as many threads as possible.
+    if (cost <= lowest_cost) {
+      tile_long_side_len = proposed_tile_long_side_len;
+      tile_short_side_len = tile_size_pair.second;
+      lowest_cost = cost;
+    }
+  }
+
+  // Request tile sizes such that the longer side of threadblock aligns with
+  // the longer side of input data block to maximize read throughput.
+  // The ideal tile shape to request is one with its length of the shorter
+  // side of the tile being equal to the length of the shorter side of the
+  // input matrix.
+  int requested_tile_size_i = input_dims[1] >= kMinDimensionToUseTiles
+                                  ? tile_long_side_len
+                                  : input_dims[1];
+  int requested_tile_size_j = input_dims[1] >= kMinDimensionToUseTiles
+                                  ? input_dims[2]
+                                  : tile_long_side_len;
+
+  // Truncate the shorter size requested according to the manual limit set in
+  // tile_spec to make sure that we do not launch configurations violating
+  // hardware limits.
+  requested_tile_size_i = requested_tile_size_i == tile_long_side_len
+                              ? tile_long_side_len
+                              : min(requested_tile_size_i, tile_short_side_len);
+  requested_tile_size_j = requested_tile_size_j == tile_long_side_len
+                              ? tile_long_side_len
+                              : min(requested_tile_size_j, tile_short_side_len);
+
+  Dimension<3> input_dims_in_tiles = {
+      input_dims[0],
+      (input_dims[1] + requested_tile_size_i - 1) / requested_tile_size_i,
+      (input_dims[2] + requested_tile_size_j - 1) / requested_tile_size_j,
+  };
+
+  int total_tiles_count =
+      input_dims_in_tiles[0] * input_dims_in_tiles[1] * input_dims_in_tiles[2];
+
+#define DISPATCH_WITH_UINT_TYPE(TYPE_SIZE)                                \
+  BatchNarrowMatrixTransposeDispatcher<uint##TYPE_SIZE, 32, 2>::DoIt(     \
+      d, requested_tile_size_i, requested_tile_size_j, total_tiles_count, \
+      reinterpret_cast<const uint##TYPE_SIZE*>(input), input_dims,        \
+      reinterpret_cast<uint##TYPE_SIZE*>(output))
+
+  switch (sizeof(T)) {
+    case 1:
+      DISPATCH_WITH_UINT_TYPE(8);
+      break;
+    case 2:
+      DISPATCH_WITH_UINT_TYPE(16);
+      break;
+    case 4:
+      DISPATCH_WITH_UINT_TYPE(32);
+      break;
+    case 8:
+      DISPATCH_WITH_UINT_TYPE(64);
+      break;
+    case 16:
+      BatchNarrowMatrixTransposeDispatcher<float4, 32, 2>::DoIt(
+          d, requested_tile_size_i, requested_tile_size_j, total_tiles_count,
+          reinterpret_cast<const float4*>(input), input_dims,
+          reinterpret_cast<float4*>(output));
+      break;
```

Comment:
I think we should static-assert that `alignof(T) >= alignof(uintXX / float4)`.  Otherwise it seems we could get into trouble.